model/fn-execution/src/main/proto/beam_interactive_api.proto

```diff
@@ -0,0 +1,106 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/*
+ * Protocol Buffers describing a service that can be used in conjunction with
+ * the TestStream class in order to control a pipeline remotely.
+ */
+
+syntax = "proto3";
+
+package org.apache.beam.model.fn_execution.v1;
+
+option go_package = "fnexecution_v1";
+option java_package = "org.apache.beam.model.fnexecution.v1";
+option java_outer_classname = "BeamInteractiveApi";
+
+import "beam_runner_api.proto";
+import "google/protobuf/timestamp.proto";
+
+service InteractiveService {
+
+  // A TestStream will request for events using this RPC.
+  rpc Events(EventsRequest) returns (stream EventsResponse) {}
+
+  rpc Start (StartRequest) returns (StartResponse) {}
```

Comment:
Why do we need a GRPC service for this? My understanding was that TestStream could be backed by a simple Grpc service that just streamed TestStreamPayload.Event messages.

Reply:
We need a GRPC to simultaneously allow for streaming of events to the TestStream and for user interactivity in a portable way. When the TestStream is implemented in the FnApiRunner, there will be a hard container boundary to both the runner and worker that needs to be crossed with some IPC. GRPC is the best way for that, currently.

Reply:
Sorry for not being more clear. 

Currently, we have a TestStream that reads its elements from memory. We are hoping to augment that so that its elements can be supplied externally, over GRPC, so we have

TestStreamImpl <--- TestStreamPayload.Event ---- Driver

Where TestStreamImpl and Driver sit in separate processes. To do this, we should augment TestStreamPayload with an (optional) address (you can probably reuse Endpoint) to which it should connect over GRPC to get a stream of TestStreamPayload.Events. 

This Service seems to be trying to split Driver across processes, which seems to introduce unnecessary complexity.

Reply:
Thanks for your clarification. We are actually on the same page. To clarify my position, the service will be sitting in the InteractiveRunner and the TestStream will ask for events from it using the EventsRequest. A user can then start replaying the stream using the StartRequest. At a later time, the user can also pause or prematurely stop the replay.

I added an endpoint to the TestStreamPayload.